src/parser.cpp

```diff
@@ -1015,11 +1021,15 @@ namespace ada::parser {
           else if (pointer == pointer_end || *pointer == '/' || *pointer == '?' ||
                                 (url.is_special() && *pointer == '\\') ||
                                 state_override.has_value()) {
+            port_end_pointer = pointer;
+
             // If buffer is not the empty string, then:
-            if (!buffer.empty()) {
+            if (*port_start_pointer != pointer) {
+              std::string _buffer = std::string(*port_start_pointer, *port_end_pointer - *port_start_pointer);
+
               // Let port be the mathematical integer value that is represented by buffer in radix-10
               // using ASCII digits for digits with values 0 through 9.
-              long port = std::atol(buffer.c_str());
+              long port = std::atol(_buffer.c_str());
```

Comment:
What is the correct way of detecting uint32_t overflow?

Reply:
If you know that the string is all made of digits, then you can just check the string length. The maximal port value is 65535, so the string should not exceed 5 characters. Best thing might be to just checking the string prior to parsing...

Reply:
But the C++ standard way would be this...

```C++

uint32_t get_port_fast(std::string_view::iterator begin, std::string_view::iterator end) noexcept {
    uint32_t port;
    std::string_view view(begin, end-begin);
    auto r = std::from_chars(view.data(), view.data() + view.size() , port);
    if(r.ec != std::errc()) { return -1; }
    return port;
}
```

`r.ec` can take different values depending on the nature of the error.

You can also return an optional from the above function if you'd like.

Reply:
In all cases, if you go the  std::string_view, std::from_chars route, you get no uninlined function call and no allocation under GCC. You get something performant.  From what I can see, in general, GCC is unable to elide the temporary string, it seems to always create it... meaning that it allocates memory (possibly on stack) and then copy your content to it, then it has to later clean it up. It could also possibly throw.

Reply:
Updated the pull request. Used std::optional, since output is unsigned integer, `-1` is not available.